src/ray/common/id.cc

```diff
@@ -93,32 +196,101 @@ uint64_t MurmurHash64A(const void *key, int len, unsigned int seed) {
   return h;
 }
 
+ActorID ActorID::FromRandom(const JobID &job_id) {
+  std::string data(kUniqueBytesLength, 0);
+  FillRandom(&data);
+  std::copy_n(job_id.Data(), JobID::kLength, std::back_inserter(data));
+  RAY_CHECK(data.size() == kLength);
+  return ActorID::FromBinary(data);
+}
+
+JobID ActorID::JobId() const {
+  RAY_CHECK(!IsNil());
+  return JobID::FromBinary(
+      std::string(reinterpret_cast<const char *>(this->Data() + kUniqueBytesLength), JobID::kLength));
+}
+
+TaskID TaskID::FromRandom() {
+  return FromRandom(ActorID::Nil());
+}
+
+TaskID TaskID::FromRandom(const ActorID &actor_id) {
+  std::string data(kUniqueBytesLength, 0);
+  FillRandom(&data);
+  std::copy_n(actor_id.Data(), ActorID::kLength, std::back_inserter(data));
+  return TaskID::FromBinary(data);
+}
+
+ActorID TaskID::ActorId() const {
+  return ActorID::FromBinary(
+      std::string(reinterpret_cast<const char *>(id_ + kUniqueBytesLength), ActorID::Size()));
+}
+
 TaskID TaskID::ComputeDriverTaskId(const WorkerID &driver_id) {
   std::string driver_id_str = driver_id.Binary();
   driver_id_str.resize(Size());
   return TaskID::FromBinary(driver_id_str);
 }
 
 TaskID ObjectID::TaskId() const {
+  if (!CreatedByTask()) {
+    // TODO(qwang): Should be RAY_CHECK here.
+    RAY_LOG(WARNING) << "Shouldn't call this on a non-task object id: " << this->Hex();
+  }
   return TaskID::FromBinary(
       std::string(reinterpret_cast<const char *>(id_), TaskID::Size()));
 }
 
-ObjectID ObjectID::ForPut(const TaskID &task_id, int64_t put_index) {
-  RAY_CHECK(put_index >= 1 && put_index <= kMaxTaskPuts) << "index=" << put_index;
-  ObjectID object_id;
-  std::memcpy(object_id.id_, task_id.Binary().c_str(), task_id.Size());
-  object_id.index_ = -put_index;
-  return object_id;
+ObjectID ObjectID::ForPut(const TaskID &task_id, ObjectIDIndexType put_index) {
+  RAY_CHECK(put_index >= 1 && put_index <= kMaxObjectIndex) << "index=" << put_index;
+
+  ObjectIDFlagsType flags = 0x0000;
+  SetCreatedByTaskFlag(true, &flags);
+  SetObjectTypeFlag(ObjectType::PUT_OBJECT, &flags);
+
+  // Use a default transport_type 0.
+  SetTransportTypeFlag(0, &flags);
```

Comment:
```suggestion
  SetTransportTypeFlag(transport_type, &flags);
```
Change the signature of the method so that the caller must pass in `transport_type`. There should be no notion of a default transport type.